### ISSUE TITLE
`required: false` prevents all other validators from running, even when attribute has non-null value

### ISSUE THREAD

**Issue by cspotcode** [balderdashy/waterline#867]
When I set `required: false` in a model attribute's configuration, none of the other validations run, even when I am setting the attribute to a non-null, non-undefined value.

For example:
```
// here's your model config for User
module.exports = {
    //...
    attributes: {
        secondaryEmail: {
            required: false,
            email: true // this validation will not execute
        }
    }
    //...
}
 
// and you try to do this
User.update({id: userId}, {
    secondaryEmail: 'not an email' // should be rejected by "email" validator
}).exec(function(err, updated) {
    assert(err == null); // the update succeeded.  Is this intended behavior?
    //...
});
```

Here is the relevant waterline source code:
https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/validations.js#L132-141

Is this intended behavior or a bug?

**Comment by tjwebb** [comment]
I'm not sure; it sounds confusing but plausible. Try using `notNull: true` and see if that fixes your issue.

**Comment by cspotcode** [comment]
Omitting required will allow other validations to run, but my question is why does Waterline have code to specifically *prevent* other validations from running when required is set to a falsey value?  And if the behavior is intentional, where is it documented?

Here is the commit that causes this behavior: https://github.com/balderdashy/waterline/commit/57afcc4de4689db813fec83d810250c4e0bea982

Based on the commit message, the author's intent was to merely allow `required: false` to mean "not required" in the same way that omitting required means "not required."  Otherwise, passing `required: false` into anchor is the same as passing `required: true`.  I'm also not sure why anchor implements that behavior, but that's a separate issue.

**Comment by tjwebb** [comment]
Yea you're right, it shouldn't just `return cb()` right there. This is basically a bug, but I'm going to rely on someone in the community to submit a PR. Since you've dug into the code already, it'd be awesome if you could submit a patch.

**Comment by cspotcode** [comment]
Just wanted to update this issue to say I don't realistically have time to submit a patch.  Hopefully someone else will get around to it.

**Comment by devinivy** [comment]
See https://github.com/balderdashy/waterline/commit/57afcc4de4689db813fec83d810250c4e0bea982#comments for info on that breaking commit from its author.

**Comment by dmarcelino** [comment]
This definitely looks wrong...
``` javascript
    // if required is set to 'false', don't enforce as required rule
    if (curValidation.hasOwnProperty('required')&&!curValidation.required) {
        return cb();
    }
```

According to @stensrud:

> ```javascript
>     // If value is not required and empty then don't
>     // try and validate it
>     if(!curValidation.required) {
>       if(value === null || value === '') return cb();
>     }
> ```
> 
> This code, present just above my commit, have the same intention. Tried to remove my commit, and it seems to work now.

So I think we can safely remove the offending lines of code and add a test to prevent regressions.

**Comment by stensrud** [comment]
Yes, my sad contribution can definitely be removed. I do still, however, believe the validation of non-required values has some problems (I submitted this code for a reason).

**Comment by dmarcelino** [comment]
Worry not, we've all been there! :smiley: 

We should add a few tests to cover all possible cases and make sure this gets fixed and remains fixed.

**Comment by dmarcelino** [comment]
Issued #982 which should fix this, at least the tests pass: 1bbba8f7148b7c9506454545944c77463743f69a! :smiley:
